lua/plenary/async_lib/async.lua

```diff
@@ -0,0 +1,217 @@
+    assert(success, "Coroutine failed")
+
+    if co.status(thread) == "dead" then
+      idle:stop()
+      callback()
+    end
+  end)
+end
+
+-- use with wrap
+local execute = function(future, callback)
+  assert(type(future) == "function", "type error :: expected func")
+  local thread = co.create(future)
+
+  local step
+  step = function(...)
+    local res = {co.resume(thread, ...)}
+    local stat = res[1]
+    local ret = {select(2, unpack(res))}
+
+    assert(stat, string.format("The coroutine failed with this message: %s", ret[1]))
+
+    if co.status(thread) == "dead" then
+      (callback or function() end)(unpack(ret))
+    else
+      assert(#ret == 1, "expected a single return value")
+      local returned_future = ret[1]
+      assert(type(returned_future) == "function", "type error :: expected func")
+      returned_future(step)
+    end
+  end
+
+  step()
+end
+
+-- use with CPS function, creates future factory
+-- must have argc for arity checking
+M.wrap = function(func, argc)
+  assert(type(func) == "function", "type error :: expected func, got " .. type(func))
+  assert(type(argc) == "number" or argc == "vararg", "expected argc to be a number or string literal 'vararg'")
+
+  return function(...)
+    local params = {...}
+
+    local function future(step)
+      if step then
+        if type(argc) == "number" then
+          params[argc] = step
+        else
+          table.insert(params, step) -- change once not optional
+        end
+        return func(unpack(params))
+      else
+        return co.yield(future)
+      end
+    end
+    return future
+  end
+end
+
+--- WIP
+local thread_loop_async = M.wrap(thread_loop, 2)
+
+-- many futures -> single future
+M.join = M.wrap(function(futures, step)
+  local len = #futures
+  local results = {}
+  local done = 0
+
+  if len == 0 then
+    return step(results)
+  end
+
+  for i, future in ipairs(futures) do
+    assert(type(future) == "function", "type error :: future must be function")
+
+    local callback = function(...)
+      results[i] = {...} -- should we set this to a table
+      done = done + 1
+      if done == len then
+        -- step(unpack(results))
+        step(results) -- should we unpack?
+      end
+    end
+
+    future(callback)
+  end
+end, 2)
+
+M.select = M.wrap(function(futures, step)
+  local selected = false
+
+  for _, future in ipairs(futures) do
+    assert(type(future) == "function", "type error :: future must be function")
+
+    local callback = function(...)
+      if not selected then
+        selected = true
+        step(...)
+      end
+    end
+
+    future(callback)
+  end
+end, 2)
+
+--- use this over running a future by calling it with no callback argument because it is more explicit
+M.run = function(future, callback)
+  future(callback or function() end)
+end
+
+M.run_all = function(futures, callback)
+  M.run(M.join(futures), callback)
+end
+
+M.await = function(future)
+  return future(nil)
```

Comment:
It seems a bit weird that here we pass `nil` and above (line 119) we pass `callback or function() end` -- I'd like to make it more consistent perhaps. Seems fine though.

Reply:
They are supposed to be completely opposite functions which is why they are not consistent. `await` will yield the function and run a future asynchronously and then resume when it is done. Passing in nil to any future will await it in an async context. `callback or function() end` is used because passing a function into a future will run it in a none async function meaning it does not yield. Not passing in a function will still run it, hence the `or`. Using the or basically forces it to be a function. Futures are created using the [`wrap`](https://github.com/nvim-lua/plenary.nvim/pull/83/files#diff-efd9f708a3e284d8c96cf671f2c63ce81766f3bf69b40e18a1c560ce3d43d887R48) function and [`async`](https://github.com/nvim-lua/plenary.nvim/pull/83/files#diff-efd9f708a3e284d8c96cf671f2c63ce81766f3bf69b40e18a1c560ce3d43d887R137) function. In those functions you can see that there are if statements checking if we passed in `nil` or a function. What type of argument we passed in controls the behavior.